feat(gateway): allow extra clients and public routes (#871)
it-at-m/refarch: docs/gateway.md

cookies.
The gateway then maps the session cookie to a JWT before routing it.

> Beside the default behavior there are special public and clients routes described in the following.

Reviewer: ```suggestion
Beside the default behavior there are special public and client routes described in the following.
```

Author: I called it `clients` everywhere as the current prefix is `/clients/**`. At least your suggestion to change the pattern would be a breaking change

#### Public Routes

> Requests matching public routes are routed WITHOUT authentication.

Reviewer: ```suggestion
Requests matching public routes are routed **WITHOUT** authentication.
```

> To configure public routes either the prefix `/public/**` can be used or custom patterns can be configured via `refarch.security.public-patterns` (see [Configuration](#configuration)).

Reviewer: ```suggestion
To configure public routes either the pre-defined prefix `/public/**` can be used or custom patterns can be configured via `refarch.security.public-patterns` (see [Configuration](#configuration)).
```

CSRF protection remains enabled for these routes and can be disabled via `refarch.security.csrf-whitelisted`.

> #### Clients Routes

Reviewer: ```suggestion
#### Client Routes
```

> Requests matching clients routes use JWT for authentication instead of session cookies.

Reviewer: ```suggestion
Requests matching client routes use JWT for authentication instead of session cookies.
```

> To configure clients routes either the prefix `/clients/**` can be used or custom patterns can be configured via `refarch.security.clients-patterns` (see [Configuration](#configuration)).

Reviewer: I'd prefer singular `client` here.

```suggestion
To configure client routes either the pre-defined prefix `/client/**` can be used or custom patterns can be configured via `refarch.security.client-patterns` (see [Configuration](#configuration)).
```

Author: ```suggestion
To configure client routes either the pre-defined prefix `/clients/**` can be used or custom patterns can be configured via `refarch.security.client-patterns` (see [Configuration](#configuration)).
```

  security:
    csrf-whitelisted: # List of routes to disable CSRF protection for (optional)
      - /example/**
    public-patterns: # Additional public routes with explicit methods (optional)

>       - pattern: /api/backend/public/**

Reviewer: Just to understand this. Wouldn't `/api/backend` resolve to backend service and pass `/public/` on to the backend? Because this would be wrong.

        methods: [GET] # All methods must be listed, there is no default all

>     client-patterns: # Additional clients routes (optional)

Reviewer: ```suggestion
    client-patterns: # Additional client routes (optional)
```

>     clients-patterns: # Additional clients routes (optional)

Reviewer: So methods can't be configured here?

```suggestion
    client-patterns: # Additional client routes (optional)
```

>       - /api/backend/clients/**

Reviewer: Just to understand this. Wouldn't `/api/backend` resolve to backend service and pass `/client/` on to the backend? Because this would be wrong.

```suggestion
      - /api/backend/client/**
```

>       - /api/backend/clients/**

Author: ```suggestion
    public-patterns: # Additional public routes with explicit methods (optional)
      - pattern: /example/**
        methods: [GET] # All methods must be listed, there is no default all
    clients-patterns: # Additional clients routes (optional)
      - /example/**
```
